Update CIL build instructions in developer guide (#2286)
TomographicImaging/CIL: docs/source/developer_guide.rst

*****************

CIL is an Object Orientated software. It has evolved during the years and it currently does not fully adhere to the following conventions. New additions must comply with
the conventions and documentation guidelines described in this section.

Building CIL from source code
==============================

Getting the code
^^^^^^^^^^^^^^^^

In case of local development and testing it is useful to be able to build the software directly. 
You should first clone this repository as:

.. code:: sh

   git clone git@github.com:TomographicImaging/CIL

The parameter ``--depth 1`` can be added to create a shallow clone with a history truncated to the specified number of commits reducing the size of the clone.
See `git documentation <https://git-scm.com/docs/git-clone#Documentation/git-clone.txt---depthdepth>`_

Building with ``pip``
^^^^^^^^^^^^^^^^^^^^^

Install Dependencies
""""""""""""""""""

We suggest creating a conda environment with all the dependencies for building CIL using the appropriate command for your operating system:

.. list-table::
   :header-rows: 1
   :widths: 20 65 15

   * - OS
     - Command
     - Status
   * - Linux

>      - ``conda env create -f https://tomography.stfc.ac.uk/scripts/env/cil_development.yml``

Reviewer: This script doesn't appear to exist?

Reviewer 2: https://github.com/TomographicImaging/CIL/pull/2286#discussion_r2788688020

     - Tested
   * - Windows

>      - ``conda env create -f https://tomography.stfc.ac.uk/scripts/env/cil_development.yml``

Reviewer: Still doesn't exist, but I would assume windows and Linux would differ.

Reviewer 2: nope, they're the same now 😄

     - Tested
   * - MacOS (ARM)
     - ``conda env create -f https://tomography.stfc.ac.uk/scripts/env/cil_development_osx.yml``
     - Experimental

.. note::
   Currently only Linux and Windows are tested and supported. The support on MacOS (ARM) is experimental and certain features are not available and not working, such as FFT filtering for FDK.

Build CIL
""""""""

A C++ compiler is required to build the source code. Let's suppose that the user is in the source directory, then the following commands should work:

.. code:: sh

   pip install -e .

.. note::
   You need to have a **working compiler** on your system, such as Visual Studio on Windows, GCC on Linux and XCode on MacOS.

If not installing inside a conda environment, then the user might need to set the locations of optional libraries:

.. code:: sh

   pip install -e . -Ccmake.define.IPP_ROOT="<path_to_ipp>" -Ccmake.define.OpenMP_ROOT="<path_to_openmp>" -Ccmake.define.CMAKE_BUILD_TYPE=RelWithDebInfo

Notes for Windows users
"""""""""""""""""""""""

> One option for development on Windows is using `WSL <https://learn.microsoft.com/en-us/windows/wsl/install#change-the-default-linux-distribution-installed>`_

Reviewer: ```suggestion
One option for development on Windows is using `Windows Subsystem for Linux (WSL) <https://learn.microsoft.com/en-us/windows/wsl/install#change-the-default-linux-distribution-installed>`_
```

Launch WSL and install build-essential using:

.. code:: sh

>    apt install build-essential

Reviewer: This isn't showing in the rendered docs

Reviewer 3: ```suggestion
.. code:: sh

   apt install build-essential
```

This will enable you to then follow the linux instructions for creating the environment and building CIL.

> Alternatively, to use Windows itself...

Reviewer: ```suggestion
```

> Install Visual Studio Community (or higher) and select the **Desktop development with C++** workload.

Reviewer: ```suggestion
To build for windows you can install Visual Studio Community (or higher) and select the **Desktop development with C++** workload.
```

.. code:: sh

   pip install -e . -Ccmake.define.IPP_ROOT="<path_to_ipp>" -Ccmake.define.OpenMP_ROOT="<path_to_openmp>" -Ccmake.define.CMAKE_BUILD_TYPE=RelWithDebInfo

Notes for Windows users
""""""""""""""""

Install Visual Studio Community (or higher) and select the **Desktop development with C++** workload.

If you are developing on Windows with conda, you need to have access to both the Visual Studio compiler and the conda environment. 

You can achieve this in two ways:
1. by opening a "Developer Command Prompt for Visual Studio" and activating the conda environment from there. This requires you 
to know the path to the `conda.bat` file, which is typically located in the `condabin` subdirectory of your conda installation. 
Once located you need to run `<path_to>\conda.bat activate <env_name>` to activate the conda environment, and then you can run the build command from there.
2. by opening the conda prompt and running the `vcvarsall.bat x64` file from the Visual Studio installation (with `x64` argument). 
This requires you to know the path to the `vcvarsall.bat` file, 

> which is typically located in the `VC/Auxiliary/Build` subdirectory of your Visual Studio installation.

Reviewer 4: ```suggestion
You can achieve this in two ways:

1. by opening a "Developer Command Prompt for Visual Studio" and activating the conda environment from there. This requires you 
   to know the path to the ``conda.bat`` file, which is typically located in the ``condabin`` subdirectory of your conda installation. 
   Once located you need to run ``<path_to>\conda.bat activate <env_name>`` to activate the conda environment, and then you can run the build command from there.
2. by opening the conda prompt and running the ``vcvarsall.bat x64`` file from the Visual Studio installation (with ``x64`` argument). 
   This requires you to know the path to the ``vcvarsall.bat`` file, 
   which is typically located in the ``VC/Auxiliary/Build`` subdirectory of your Visual Studio installation.
```

The documentation was not rendering the bullets correctly, tiny fix for that.

Along with that I also added an extra backtick to all of the backticked text to make it show up as code blocks.

     - Tested
   * - Windows
     - ``conda env create -f https://tomography.stfc.ac.uk/scripts/env/cil_development.yml``
     - Tested
   * - MacOS (ARM)
     - ``conda env create -f https://tomography.stfc.ac.uk/scripts/env/cil_development_osx.yml``
     - Experimental

.. note::
   Currently only Linux and Windows are tested and supported. The support on MacOS (ARM) is experimental and certain features are not available and not working, such as FFT filtering for FDK.

Build CIL
""""""""

A C++ compiler is required to build the source code. Let's suppose that the user is in the source directory, then the following commands should work:

.. list-table::
   :header-rows: 1
   :widths: 20 80

.. code::sh

   pip install -e .

.. note::
   You need to have a **working compiler** on your system, such as Visual Studio on Windows, GCC on Linux and XCode on MacOS.

If not installing inside a conda environment, then the user might need to set the locations of optional libraries:

.. code:: sh

   pip install -e . -Ccmake.define.IPP_ROOT="<path_to_ipp>" -Ccmake.define.OpenMP_ROOT="<path_to_openmp>" -Ccmake.define.CMAKE_BUILD_TYPE=RelWithDebInfo

Notes for Windows users
""""""""""""""""

> 

Reviewer 2: ```suggestion
One option for development on Windows is using [WSL](:https://learn.microsoft.com/en-us/windows/wsl/install#change-the-default-linux-distribution-installed).
Launch WSL and install build-essential using:
``apt install build-essentials``
This will enable you to then follow the linux instructions for creating the environment and building CIL.

Alternatively, to use Windows itself...

```

Install Visual Studio Community (or higher) and select the **Desktop development with C++** workload.

> If you are developing on Windows with conda, you need to have access to both the Visual Studio compiler and the conda environment. 

Reviewer 2: ```suggestion
If you are developing on Windows with conda, you need to have access to both the Visual Studio compiler and have created the conda environment using the command for Windows above. 
```

> You can achieve this in two ways:

Reviewer 2: I just retested and found the instructions below worked for me, so I am happy that they are correct provided we state versions we tested with. I added the version of Visual Studio I had, as a suggested change Any other versions we should include a note of?

> 1. by opening a "Developer Command Prompt for Visual Studio" and activating the conda environment from there. This requires you 

Reviewer 2: ```suggestion
1. by opening a "Developer Command Prompt for VS" and activating the conda environment from there. This requires you 
```

I do not have anything called "Developer Command Prompt for Visual Studio"

Reviewer 2: Please can you check the name of yours @paskino , I wonder if its a typo or if the name of the program is different in different versions

Reviewer: `x64 Native Tools Command Prompt for VS`
 might be the better option here?

Reviewer 2: yes that's what mine is called, and was in my original instructions

   to know the path to the ``conda.bat`` file, which is typically located in the ``condabin`` subdirectory of your conda installation. 
   Once located you need to run ``<path_to>\conda.bat activate <env_name>`` to activate the conda environment, and then you can run the build command from there.

> 2. by opening the conda prompt and running the ``vcvarsall.bat x64`` file from the Visual Studio installation (with ``x64`` argument). 

Reviewer 2: ```suggestion
2. by opening the conda prompt and running the ``vcvarsall.bat x64`` file from the Visual Studio installation (with ``x64`` argument), and then running the build command.
```

   This requires you to know the path to the ``vcvarsall.bat`` file, 
   which is typically located in the ``VC/Auxiliary/Build`` subdirectory of your Visual Studio installation.

> 

Reviewer 2: ```suggestion

Note: we tested these instructions with Visual Studio 2026 version 18.1.1
```

.. automethod:: cil.recon.FBP.FBP.run

Building CIL from source code
-----------------------------

Getting the code
^^^^^^^^^^^^^^^^

In case of local development and testing it is useful to be able to build the software directly. 
You should first clone this repository as

.. code:: sh

   git clone git@github.com:TomographicImaging/CIL

Building with ``pip``
^^^^^^^^^^^^^^^^^^^^^

Install Dependencies
""""""""""""""""""

We suggest creating a conda environment with all the dependencies for building CIL running the following shell script depending on operating system:

.. list-table::
   :header-rows: 1
   :widths: 20 65 15

   * - OS
     - Command
     - Status
   * - Linux

>      - ``conda env create -f https://tomographicimaging.github.io/scripts/env/cil_development.yml``

Reviewer 2: For anyone else who is testing the files are here and you need to use them instead of the url:

https://github.com/TomographicImaging/scripts/pull/9/changes

>      - ``conda env create -f https://tomographicimaging.github.io/scripts/env/cil_development.yml``

Author: That's right, I didn't want to link to a PR, so I wrote the link as it will be once [#9](https://github.com/TomographicImaging/scripts/pull/9) is merged

     - Tested
   * - Windows
     - ``conda env create -f https://tomographicimaging.github.io/scripts/env/cil_development.yml``
     - Tested
   * - MacOS (ARM)

>      - ``conda env create -f https://tomographicimaging.github.io/scripts/env/cil_development_osx.yml``

Reviewer 3: ```suggestion
     - ``conda env create -f https://tomography.stfc.ac.uk/scripts/env/cil_development.yml``
     - Tested
   * - Windows
     - ``conda env create -f https://tomography.stfc.ac.uk/scripts/env/cil_development.yml``
     - Tested
   * - MacOS (ARM)
     - ``conda env create -f https://tomography.stfc.ac.uk/scripts/env/cil_development_osx.yml``
```

Author: I like this, currently they return a 404.

     - Experimental

.. note::
   Currently only Linux and Windows are tested and supported. The support on MacOS (ARM) is experimental and certain features are not available and not working, such as FFT filtering for FDK.

Build CIL
""""""""

A C++ compiler is required to build the source code. Let's suppose that the user is in the source directory, then the following commands should work:

.. list-table::
   :header-rows: 1
   :widths: 20 80

   * - OS
     - Build Command
   * - Linux

>      - ``pip install --no-deps . -Ccmake.define.IPP_ROOT=$CONDA_PREFIX``

Reviewer 2: I followed the instructions for linux and got:

ModuleNotFoundError: No module named 'cil'

Author: I just tested with an environment where I install cmake and scikit-build from conda rather than let it handle by pip and: 
1. `pip install --no-deps .` is sufficient, i.e. IPP is found
2. `cilacc` seems to be built all the times you issue `pip install`

Reviewer 2: I don't think this matters that much anyway because running with -e:
pip install --no-deps . -Ccmake.define.IPP_ROOT=$CONDA_PREFIX
means you only have to pip install once anyway

Reviewer 3: resolving this as handled by `FindIPP.cmake`

   * - Windows

>      - ``pip install --no-deps . -Ccmake.define.IPP_ROOT=%CONDA_PREFIX%``

Reviewer 3: IPP_ROOT shouldn't be needed in a conda environment.

```suggestion
.. code::sh

   pip install -e .
```
